File: computeCirc_v1.1/src/main.f
! Compile with pgf90


      program computeC

      implicit none

      include '../include/const.h'
	
      integer i, j, k, n, nx, ny, nz, nv
      integer nfiles, istatus

      real, allocatable :: x(:)
      real, allocatable :: y(:)
      real, allocatable :: z(:)
      real, allocatable :: u(:,:,:)
      real, allocatable :: v(:,:,:)
      real, allocatable :: circ(:,:,:)

      real dx, dy, dz

      character*100 infile(maxfiles)
      character*8 label
      character*8 u_variable, v_variable
      real Cscale, radius

      namelist /inputparms/ nfiles, infile, u_variable,
     >                      v_variable, radius

!-----------------------------------------------------------------      
! Scale factors

      Cscale = 1.e-4

!-----------------------------------------------------------------      
! Obtain input parameters

      read(5, inputparms)

!-----------------------------------------------------------------
! Begin main program loop

      do n = 1, nfiles

! Determine size of arrays and get data from netcdf file.

        print*,'Computing circulation...'

        call getsize(infile(n),nx,ny,nz,nv)
        allocate (x(nx),y(ny),z(nz))
        call getgridinfo(infile(n),dx,dy,dz,x,y,z,nx,ny,nz)

        allocate ( u(nx,ny,nz), v(nx,ny,nz), circ(nx,ny,nz) ) 

        call netcdf_read(u,u_variable,infile(n),nx,ny,nz)
        call netcdf_read(v,v_variable,infile(n),nx,ny,nz)

        call getcirc(u,v,x,y,z,dx,dy,dz,nx,ny,nz,radius,circ)

        do i = 1, nx
        do j = 1, ny
        do k = 1, nz

           if ( circ(i,j,k) .ne. missing_val ) then
             circ(i,j,k) = circ(i,j,k)*Cscale
           endif

        enddo
        enddo 
        enddo


        print*,'Done.'
        print*, ' '
        print*,'Writing data to netcdf file...'

        label = 'CIRC    '
        call varinq(label,infile(n),istatus)
        if ( istatus .eq. 0 ) then
           call netcdf_write(circ,label,infile(n),nx,ny,nz)
        else
           call netcdf_overwrite(circ,label,infile(n),nx,ny,nz)
        endif

        print*,'Done.'
        print*, ' '
        print*,'Operation completed for ', infile(n)
        print*, ' ' 
        print*, ' '

! End of main program loop

        deallocate (u,v,circ,x,y,z)

      enddo      

!------------------------------------------------------------------

      end


!    END OF MAIN CODE
!------------------------------------------------------------------


      
 

File: computeCirc_v1.1/src/main.py
#!/usr/bin/env python
"""Calculate circulation for given files.

Read list of files and radius of circles for circulation calculations
from standard input.
"""
import ast
import configparser
import io
import re
import sys

import numpy as np

from . import circ
from . import const
from . import netcdfio

C_SCALE = 1.0e-4


def parse_namelist(fileptr: "io.TextIOBase") -> configparser.ConfigParser:
    """Parse a namelist.

    Parameters
    ----------
    fileptr : file
        A file containing a namelist

    Returns
    -------
    configparser.ConfigParser
        The parsed namelist

    Examples
    --------
    FIXME: Add docs.

    See Also
    --------
    f20nml
        An alternate and more robust namelist parser.
    """
    namelist_parser = configparser.ConfigParser(
        inline_comment_prefixes="!",
        comment_prefixes="!",
        converters={
            "int": lambda val: int(val.rstrip(",")),
            "float": lambda val: float(val.rstrip(",")),
            "string": lambda val: ast.literal_eval(val.rstrip(",")),
        },
    )
    namelist_parser.SECTCRE = re.compile(r"\s*&(?P<header>\w+)\s*")

    namelist_parser.read_file(fileptr)
    return namelist_parser

def compute_circulation(fileptr: io.TextIOBase):
    namelist_parser = parse_namelist(fileptr)
    inputparms = namelist_parser["inputparms"]
    n_files = inputparms.getint("nfiles")
    u_variable = inputparms.getstring("u_variable")
    v_variable = inputparms.getstring("v_variable")
    radius = inputparms.getfloat("radius")
    # will have to change this if someone uses array notation in their
    # namelist
    in_files = [
        inputparms.getstring(key)
        for key in inputparms.keys()
        if key.startswith("infile")
    ]

    for in_file_name in in_files:
        print("computing circulation...")  # noqa: t001
        nx, ny, nz, nv = netcdfio.getsize(in_file_name)
        dx, dy, dz, x, y, z = netcdfio.getgridinfo(in_file_name, nx, ny, nz)

        u = netcdfio.netcdf_read(u_variable, in_file_name, nx, ny, nz)
        v = netcdfio.netcdf_read(v_variable, in_file_name, nx, ny, nz)

        circulation = circ.getcirc(
            u[0], v[0], x, y, z, dx, dy, dz, radius, nx, ny, nz
        )

        circulation *= C_SCALE

        print("Done\n\nWriting data to netcdf file")  # noqa: T001

        label = "CIRC"
        istatus = netcdfio.varinq(label, in_file_name)
        if not istatus:
            netcdfio.netcdf_write(circulation, label, in_file_name, nx, ny, nz)
        else:
            netcdfio.netcdf_overwrite(circulation, label, in_file_name, nx, ny, nz)

        print("Done\n\nOperation completed for", in_file_name, "\n\n")  # noqa: T001


def main() -> int:
    compute_circulation(sys.stdin)
    return 0


if __name__ == "__main__":
    sys.exit(main())
